Lib/test/support/__init__.py

```diff
@@ -3034,3 +3034,39 @@ def collision_stats(nbins, nballs):
         collisions = k - occupied
         var = dn*(dn-1)*((dn-2)/dn)**k + meanempty * (1 - meanempty)
         return float(collisions), float(var.sqrt())
+
+
+class catch_unraisable_exception:
+    """
+    Context manager catching unraisable exception using sys.unraisablehook.
+
+    Usage:
+
+        try:
+            with support.catch_unraisable_exception() as cm:
+                ...
+
+            # check the expected unraisable exception
+            ...
+        finally:
```

Comment:
this try/with/finally is somewhat tricky, how about:
```python
try:
    with support.throw_unraisable_exceptions():
        ...
except Exception as e:
    ... # the exception is now here
```

eg:
```python
@contextlib.contextmanager
def throw_unraisable_exceptions():
    unraisable = None
    old_hook = sys.unraisablehook

    def hook(exc):
        nonlocal unraisable
        unraisable = exc

    sys.unraisablehook = hook
    try:
        yield
        if unraisable is not None:
            raise unraisable
    finally:
        unraisable = None
```

Reply:
or recommend usage like this:
```python
    with support.catch_unraisable_exceptions() as cm:
        ...
        cm.unraisable  # only available here
    assert cm.unraisable is None  # now it's gone
```

```
   def __exit__(self, *exc_info):
        self.unraisable = None  # clear unraisable here
        sys.unraisablehook = self._old_hook
```

Reply:
Good idea :-) I implemented your ``__exit__`` idea to avoid the need for try/finally.

throw_unraisable_exceptions() might be useful, but modified tests needs to access the 'obj' attribute of the unraisable hook. Later, they might also want to get access to the 'err_msg' attribute: https://github.com/python/cpython/pull/13488

Reply:
There's still a sharp edge here:

how about making `cm.unraisable` throw if accessed outside the context instead of being None, eg:

```
def __exit__(self, *exc_info):
    del self.unraisable
    sys.unraisablehook = self._old_hook
```

Reply:
and if you want to access additional unraisablehook info:

```python
class UnraisableException(Exception):
    pass


@contextlib.contextmanager
def throw_unraisable_exceptions():
    unraisable = None
    extra = None
    old_hook = sys.unraisablehook

    def hook(exc, *args, **kwargs):
        nonlocal unraisable, extra
        unraisable = exc
        extra = (args, kwargs)

    sys.unraisablehook = hook
    try:
        yield
        if unraisable is not None:
            args, kwargs = extra
            raise UnraisableException(*args, **kwargs) from unraisable
    finally:
        unraisable = None
        sys.unraisablehook = old_hook
```

Reply:
I wrote PR #13554 to implement your "del self.unraisable" idea.